Userland/Libraries/LibCrypto/Curves/X25519.cpp

```diff
@@ -0,0 +1,324 @@
+
+    for (auto i = 0; i < X25519::WORDS; i++) {
+        temp += (u64)first[i] * second;
+        output[i] = temp & 0xFFFFFFFF;
+        temp >>= 32;
+    }
+
+    temp *= 38;
+    temp += (output[7] >> 31) * 19;
+    output[7] &= 0x7FFFFFFF;
+
+    for (auto i = 0; i < X25519::WORDS; i++) {
+        temp += output[i];
+        output[i] = temp & 0xFFFFFFFF;
+        temp >>= 32;
+    }
+
+    modular_reduce(state, output);
+}
+
+void X25519::modular_square(u32* state, u32* value)
+{
+    // Compute R = (A ^ 2) mod p
+    modular_multiply(state, value, value);
+}
+
+void X25519::modular_multiply(u32* state, u32* first, u32* second)
+{
+    // Compute R = (A * B) mod p
+    u64 temp = 0;
+    u64 carry = 0;
+    u32 output[16];
+
+    // Comba's method
+    for (auto i = 0; i < 16; i++) {
+        if (i < X25519::WORDS) {
+            for (auto j = 0; j <= i; j++) {
+                temp += (u64)first[j] * second[i - j];
+                carry += temp >> 32;
+                temp &= 0xFFFFFFFF;
+            }
+        } else {
+            for (auto j = i - 7; j < X25519::WORDS; j++) {
+                temp += (u64)first[j] * second[i - j];
+                carry += temp >> 32;
+                temp &= 0xFFFFFFFF;
+            }
+        }
+
+        output[i] = temp & 0xFFFFFFFF;
+        temp = carry & 0xFFFFFFFF;
+        carry >>= 32;
+    }
+
+    // Reduce bit 255 (2^255 = 19 mod p)
+    temp = (output[7] >> 31) * 19;
+    // Mask the most significant bit
+    output[7] &= 0x7FFFFFFF;
+
+    // Fast modular reduction 1st pass
+    for (auto i = 0; i < X25519::WORDS; i++) {
+        temp += output[i];
+        temp += (u64)output[i + 8] * 38;
+        output[i] = temp & 0xFFFFFFFF;
+        temp >>= 32;
+    }
+
+    //Reduce bit 256 (2^256 = 38 mod p)
+    temp *= 38;
+
+    //Reduce bit 255 (2^255 = 19 mod p)
+    temp += (output[7] >> 31) * 19;
+    //Mask the most significant bit
+    output[7] &= 0x7FFFFFFF;
+
+    // Fast modular reduction 2nd pass
+    for (auto i = 0; i < X25519::WORDS; i++) {
+        temp += output[i];
+        output[i] = temp & 0xFFFFFFFF;
+        temp >>= 32;
+    }
+
+    modular_reduce(state, output);
+}
+
+void X25519::modular_add(u32* state, u32* first, u32* second)
+{
+    // R = (A + B) mod p
+    u64 temp = 0;
+    for (auto i = 0; i < X25519::WORDS; i++) {
+        temp += first[i];
+        temp += second[i];
+        state[i] = temp & 0xFFFFFFFF;
+        temp >>= 32;
+    }
+
+    modular_reduce(state, state);
+}
+
+void X25519::modular_subtract(u32* state, u32* first, u32* second)
+{
+    // R = (A - B) mod p
+    i64 temp = -19;
+    for (auto i = 0; i < X25519::WORDS; i++) {
+        temp += first[i];
+        temp -= second[i];
+        state[i] = temp & 0xFFFFFFFF;
+        temp >>= 32;
+    }
+
+    state[7] += 0x80000000;
+
+    modular_reduce(state, state);
+}
+
+void X25519::modular_reduce(u32* state, u32* data)
+{
+    // R = A mod p
+    u64 temp = 19;
+    u32 other[8];
```

Comment:
If the size of this depends on x25519::WORDS, just use that constant instead of a number literal.

Reply:
Fixed